File: ruby/GrabzIt/lib/grabzit/imageoptions.rb
module GrabzIt
	require File.join(File.dirname(__FILE__), 'baseoptions')
	
	# Represents all of the options available when creating an image
	# @version 3.0
	# @author GrabzIt
	class ImageOptions < BaseOptions
		def initialize()
			super()
			@browserWidth = nil
			@browserHeight = nil
			@width = nil
			@height = nil
			@format = nil
			@targetElement = nil
			@hideElement = nil
			@waitForElement = nil
			@requestAs = 0
			@customWaterMarkId = nil
			@quality = -1
			@transparent = false
			@noAds = false
			@noCookieNotifications = false
			@address = nil
			@hd = false
			@clickElement = nil
			@jsCode = nil
		end
		
		# @return [Integer] the width of the browser in pixels
		def browserWidth
			@browserWidth
		end
		
		# Set the width of the browser in pixels
		#
		# @param value [Integer] the browser width
		# @return [void]		
		def browserWidth=(value)
			@browserWidth = value
		end	
		
		# @return [Integer] the height of the browser in pixels
		def browserHeight
			@browserHeight
		end
		
		# Set the height of the browser in pixels. Use -1 to screenshot the whole web page
		#
		# @param value [Integer] the browser height
		# @return [void]		
		def browserHeight=(value)
			@browserHeight = value
		end			
		
		# @return [Integer] get the width of the resulting screenshot in pixels.
		def width
			@width
		end
		
		# Set the width of the resulting screenshot in pixels. Use -1 to not reduce the width of the screenshot
		#
		# @param value [Integer] the width
		# @return [void]		
		def width=(value)
			@width = value
		end			
		
		# @return [Integer] get the height of the resulting screenshot in pixels
		def height
			@height
		end
		
		# Set the height of the resulting screenshot in pixels. Use -1 to not reduce the height of the screenshot
		#
		# @param value [Integer] the height
		# @return [void]		
		def height=(value)
			@height = value
		end

		# @return [String] get the format of the screenshot image
		def format
			@format
		end
		
		# Set the format the screenshot should be in: bmp8, bmp16, bmp24, bmp, tiff, jpg, png
		#
		# @param value [String] the format
		# @return [void]		
		def format=(value)
			@format = value
		end

		# @return [Integer] get the number of milliseconds to wait before creating the capture
		def delay
			@delay
		end
		
		# Set the number of milliseconds to wait before creating the capture
		#
		# @param value [Integer] the delay
		# @return [void]		
		def delay=(value)
			@delay = value
		end

		# @return [String] get the CSS selector of the only HTML element in the web page to capture
		def targetElement
			@targetElement
		end
		
		# Set the CSS selector of the only HTML element in the web page to capture
		#
		# @param value [String] the target element
		# @return [void]		
		def targetElement=(value)
			@targetElement = value
		end
		
		# @return [String] get the CSS selector(s) of the one or more HTML elements in the web page to hide
		def hideElement
			@hideElement
		end
		
		# Set the CSS selector(s) of the one or more HTML elements in the web page to hide
		#
		# @param value [String] the element(s) to hide
		# @return [void]		
		def hideElement=(value)
			@hideElement = value
		end
		
		# @return [String] get the CSS selector of the HTML element in the web page that must be visible before the capture is performed
		def waitForElement
			@waitForElement
		end
		
		# Set the CSS selector of the HTML element in the web page that must be visible before the capture is performed
		#
		# @param value [String] the element to wait for
		# @return [void]		
		def waitForElement=(value)
			@waitForElement = value
		end
		
		# @return [String] get the CSS selector of the HTML element in the web page that must clicked before the capture is performed
		def clickElement
			@clickElement
		end
		
		# Set the CSS selector of the HTML element in the web page that must clicked before the capture is performed
		#
		# @param value [String] the element to click
		# @return [void]		
		def clickElement=(value)
			@clickElement = value
		end		
		
		# @return [Integer] get which user agent type should be used
		def requestAs
			@requestAs
		end
		
		# Set which user agent type should be used: Standard Browser = 0, Mobile Browser = 1, Search Engine = 2 and Fallback Browser = 3
		#
		# @param value [Integer] the browser type
		# @return [void]		
		def requestAs=(value)
			@requestAs = value
		end		
		
		# @return [String] the custom watermark id.
		def customWaterMarkId
			@customWaterMarkId
		end
		
		# Set a custom watermark to add to the screenshot.
		#
		# @param value [String] custom watermark id
		# @return [void]		
		def customWaterMarkId=(value)
			@customWaterMarkId = value
		end					
		
		# @return [Integer] the quality of the screenshot.
		def quality
			@quality
		end
		
		# Set the quality of the screenshot where 0 is poor and 100 excellent. The default is -1 which uses the recommended quality
		#
		# @param value [Integer] the quality
		# @return [void]		
		def quality=(value)
			@quality = value
		end	

		# @return [Boolean] true if the image capture should be transparent
		def transparent
			@transparent
		end
		
		# Set to true if the image capture should be transparent. This is only compatible with png and tiff images
		#
		# @param value [Boolean] true if the image should be transparent
		# @return [void]
		def transparent=(value)
			@transparent = value
		end
		
		# @return [Boolean] true if the image capture should be in high definition
		def hd
			@hd
		end
		
		# Set to true if the image capture should be in high definition.
		#
		# @param value [Boolean] true if the image should be high definition
		# @return [void]
		def hd=(value)
			@hd = value
		end
		
		# @return [Boolean] get if adverts should be automatically hidden
		def noAds
			@noAds
		end
		
		# Set to true if adverts should be automatically hidden
		#
		# @param value [Boolean] hide adverts
		# @return [void]		
		def noAds=(value)
			@noAds = value
		end		

		# @return [Boolean] get if cookie notifications should be automatically hidden
		def noCookieNotifications
			@noCookieNotifications
		end
		
		# Set to true if cookie notifications should be automatically hidden
		#
		# @param value [Boolean] hide cookie notifications
		# @return [void]		
		def noCookieNotifications=(value)
			@noCookieNotifications = value
		end

		# @return [String] get the URL to execute the HTML code in
		def address
			@address
		end
		
		# Set the URL to execute the HTML code in
		#
		# @param value [String] the address
		# @return [void]		
		def address=(value)
			@address = value
		end		

		# @return [String] get the JavaScript code that will be execute in the web page before the capture is performed
		def jsCode
			@jsCode
		end
		
		# Set the JavaScript code that will be execute in the web page before the capture is performed
		#
		# @param value [String] the javascript to execute
		# @return [void]		
		def jsCode=(value)
			@jsCode = value
		end	
		
		# Define a HTTP Post parameter and optionally value, this method can be called multiple times to add multiple parameters. Using this method will force 
		# GrabzIt to perform a HTTP post.
		#
		# @param name [String] the name of the HTTP Post parameter
		# @param value [String] the value of the HTTP Post parameter
		def add_post_parameter(name, value)
			@post = appendParameter(@post, name, value)
		end		
		
		# @!visibility private
		def _getSignatureString(applicationSecret, callBackURL, url = nil)
			items = [applicationSecret]
			
			if(url != nil)
				items.push(GrabzIt::Utility.nil_check(url))
			end
			
			items.push(GrabzIt::Utility.nil_check(callBackURL),GrabzIt::Utility.nil_check(@format),GrabzIt::Utility.nil_int_check(@height),GrabzIt::Utility.nil_int_check(@width),GrabzIt::Utility.nil_int_check(@browserHeight),GrabzIt::Utility.nil_int_check(@browserWidth),GrabzIt::Utility.nil_check(@customId),GrabzIt::Utility.nil_int_check(@delay),GrabzIt::Utility.nil_check(@targetElement),GrabzIt::Utility.nil_check(@customWaterMarkId),GrabzIt::Utility.nil_int_check(@requestAs),GrabzIt::Utility.nil_check(@country),GrabzIt::Utility.nil_int_check(@quality),GrabzIt::Utility.nil_check(@hideElement),GrabzIt::Utility.nil_check(@exportURL),GrabzIt::Utility.nil_check(@waitForElement),GrabzIt::Utility.b_to_str(@transparent),GrabzIt::Utility.nil_check(@encryptionKey),GrabzIt::Utility.b_to_str(@noAds),GrabzIt::Utility.nil_check(@post),GrabzIt::Utility.nil_check(@proxy),GrabzIt::Utility.nil_check(@address),GrabzIt::Utility.b_to_str(@noCookieNotifications),GrabzIt::Utility.b_to_str(@hd),GrabzIt::Utility.nil_check(@clickElement),GrabzIt::Utility.nil_check(@jsCode))
			
			return items.join("|")
		end
	
		# @!visibility private
		def _getParameters(applicationKey, sig, callBackURL, dataName, dataValue)
			params = createParameters(applicationKey, sig, callBackURL, dataName, dataValue)		
			params['width'] = GrabzIt::Utility.nil_int_check(@width)
			params['height'] = GrabzIt::Utility.nil_int_check(@height)
			params['format'] = GrabzIt::Utility.nil_check(@format)
			params['bwidth'] = GrabzIt::Utility.nil_int_check(@browserWidth)
			params['customwatermarkid'] = GrabzIt::Utility.nil_check(@customWaterMarkId)
			params['bheight'] = GrabzIt::Utility.nil_int_check(@browserHeight)
			params['delay'] = GrabzIt::Utility.nil_int_check(@delay)
			params['target'] = GrabzIt::Utility.nil_check(@targetElement)
			params['hide'] = GrabzIt::Utility.nil_check(@hideElement)
			params['waitfor'] = GrabzIt::Utility.nil_check(@waitForElement)
			params['requestmobileversion'] = GrabzIt::Utility.nil_int_check(@requestAs)		
			params['quality'] = GrabzIt::Utility.nil_int_check(@quality)
			params['transparent'] = GrabzIt::Utility.b_to_str(@transparent)
			params['noads'] = GrabzIt::Utility.b_to_str(@noAds)
			params['post'] = GrabzIt::Utility.nil_check(@post)
			params['address'] = GrabzIt::Utility.nil_check(@address)
			params['nonotify'] = GrabzIt::Utility.b_to_str(@noCookieNotifications)
			params['hd'] = GrabzIt::Utility.b_to_str(@hd)
			params['click'] = GrabzIt::Utility.nil_check(@clickElement)
			params['jscode'] = GrabzIt::Utility.nil_check(@jsCode)
			
			return params
		end
	end
end

File: java/grabzit/src/main/java/it/grabz/grabzit/parameters/ImageOptions.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package it.grabz.grabzit.parameters;

import it.grabz.grabzit.enums.BrowserType;
import it.grabz.grabzit.enums.ImageFormat;
import java.io.UnsupportedEncodingException;
import java.util.HashMap;

/**
 * Contains all of the available options when creating a image capture.
 * 
 * @version 3.0
 * @author GrabzIt
 */
public class ImageOptions extends BaseOptions {
    private int browserWidth;
    private int browserHeight;
    private int outputWidth;
    private int outputHeight;
    private ImageFormat format;
    private String targetElement;
    private String hideElement;
    private String waitForElement;    
    private BrowserType requestAs;
    private String customWaterMarkId;
    private int quality;
    private boolean transparent;
    private boolean noAds;
    private boolean noCookieNotifications;
    private String address;
    private boolean hd;
    private String clickElement;

    public ImageOptions()
    {
        this.browserHeight = 0;
        this.browserWidth = 0;
        this.outputHeight = 0;
        this.outputWidth = 0;
        this.format = ImageFormat.JPG;
        this.targetElement = "";
        this.hideElement = "";
        this.waitForElement = "";
        this.requestAs = BrowserType.STANDARDBROWSER;
        this.customWaterMarkId = "";
        this.quality = -1;
        this.address = "";
        this.clickElement = "";
    }
    
    /**
     * @return the width of the browser in pixels.
     */
    public int getBrowserWidth() {
        return browserWidth;
    }

    /**
     * @param browserWidth set the width of the browser in pixels.
     */
    public void setBrowserWidth(int browserWidth) {
        this.browserWidth = browserWidth;
    }

    /**
     * @return the height of the browser in pixels.
     */
    public int getBrowserHeight() {
        return browserHeight;
    }

    /**
     * @param browserHeight set the height of the browser in pixels. Use -1 to screenshot the whole web page.
     */
    public void setBrowserHeight(int browserHeight) {
        this.browserHeight = browserHeight;
    }

    /**
     * @return the width of the resulting screenshot in pixels.
     */
    public int getOutputWidth() {
        return outputWidth;
    }

    /**
     * @param outputWidth set the width of the resulting screenshot in pixels. Use -1 to not reduce the width of the screenshot.
     */
    public void setOutputWidth(int outputWidth) {
        this.outputWidth = outputWidth;
    }

    /**
     * @return get the height of the resulting screenshot in pixels.
     */
    public int getOutputHeight() {
        return outputHeight;
    }

    /**
     * @param outputHeight set the height of the resulting screenshot in pixels. Use -1 to not reduce the height of the screenshot.
     */
    public void setOutputHeight(int outputHeight) {
        this.outputHeight = outputHeight;
    }

    /**
     * @return the format of the screenshot image.
     */
    public ImageFormat getFormat() {
        return format;
    }

    /**
     * @param format set the format the screenshot should be in: bmp8, bmp16, bmp24, bmp, tiff, jpg, png.
     */
    public void setFormat(ImageFormat format) {
        this.format = format;
    }

    /**
     * @return the number of milliseconds to wait before creating the capture.
     */
    public int getDelay() {
        return delay;
    }

    /**
     * @param delay set the number of milliseconds to wait before creating the capture.
     */
    public void setDelay(int delay) {
        this.delay = delay;
    }

    /**
     * @return get the CSS selector of the HTML element to click.
     */
    public String getClickElement() {
        return clickElement;
    }

    /**
     * @param clickElement set the CSS selector of the HTML element to click.
     */
    public void setClickElement(String clickElement) {
        this.clickElement = clickElement;
    }     
    
    /**
     * @return get the CSS selector of the only HTML element in the web page to capture.
     */
    public String getTargetElement() {
        return targetElement;
    }

    /**
     * @param targetElement set the CSS selector of the only HTML element in the web page to capture.
     */
    public void setTargetElement(String targetElement) {
        this.targetElement = targetElement;
    }

    /**
     * @return get the CSS selector(s) of the one or more HTML elements in the web page to hide.
     */
    public String getHideElement() {
        return hideElement;
    }

    /**
     * @param hideElement set the CSS selector(s) of the one or more HTML elements in the web page to hide.
     */
    public void setHideElement(String hideElement) {
        this.hideElement = hideElement;
    }    
    
    /**
     * @return the waitForElement the CSS selector of the HTML element in the web page that must be visible before the capture is performed.
     */
    public String getWaitForElement() {
        return waitForElement;
    }

    /**
     * @param waitForElement the CSS selector of the HTML element in the web page that must be visible before the capture is performed.
     */
    public void setWaitForElement(String waitForElement) {
        this.waitForElement = waitForElement;
    }        
        
    /**
     * @return get which user agent type should be used.
     */
    public BrowserType getRequestAs() {
        return requestAs;
    }

    /**
     * @param requestAs set which user agent type should be used.
     */
    public void setRequestAs(BrowserType requestAs) {
        this.requestAs = requestAs;
    }

    /**
     * @return the custom watermark id.
     */
    public String getCustomWaterMarkId() {
        return customWaterMarkId;
    }

    /**
     * @param customWaterMarkId set a custom watermark to add to the screenshot.
     */
    public void setCustomWaterMarkId(String customWaterMarkId) {
        this.customWaterMarkId = customWaterMarkId;
    }

    /**
     * @return the quality of the screenshot.
     */
    public int getQuality() {
        return quality;
    }

    /**
     * @param quality set the quality of the screenshot where 0 is poor and 100 excellent. The default is -1 which uses the recommended quality.
     */
    public void setQuality(int quality) {
        this.quality = quality;
    }
    
     /**
     * @return if the image capture should be transparent.
     */
    public boolean isTransparent() {
        return transparent;
    }

    /**
     * @param transparent set to true if the image capture should be transparent. This is only compatible with png and tiff images.
     */
    public void setTransparent(boolean transparent) {
        this.transparent = transparent;
    }   
    
    /**
     * @return if true adverts should be automatically hidden.
     */
    public boolean isNoAds() {
        return noAds;
    }

    /**
     * @param noAds set to true if adverts should be automatically hidden.
     */
    public void setNoAds(boolean noAds) {
        this.noAds = noAds;
    }    
    
    /**
     * @return if true cookie notifications should be automatically hidden.
     */
    public boolean isNoCookieNotifications() {
        return noCookieNotifications;
    }

    /**
     * @param noCookieNotifications set to true if cookie notification should be automatically hidden.
     */
    public void setNoCookieNotifications(boolean noCookieNotifications) {
        this.noCookieNotifications = noCookieNotifications;
    }    
    
    /**
     * @return the URL to execute the HTML code in.
     */
    public String getAddress() {
        return address;
    }

    /**
     * @param address the URL to execute the HTML code in.
     */
    public void setAddress(String address) {
        this.address = address;
    }        

    /**
     * @return true if the image should be in high definition
     */
    public boolean isHd() {
        return hd;
    }

    /**
     * @param hd set to true if the image should be in high definition
     */
    public void setHd(boolean hd) {
        this.hd = hd;
    }    
    
    /**
     * Define a HTTP Post parameter and optionally value, this method can be called multiple times to add multiple parameters. Using this method will force 
     * GrabzIt to perform a HTTP post.        
     * 
     * @param name - The name of the HTTP Post parameter
     * @param value - The value of the HTTP Post parameter
     * @throws UnsupportedEncodingException 
     */
    public void AddPostParameter(String name, String value) throws UnsupportedEncodingException
    {
        this.post = appendParameter(this.post, name, value);
    }    
    
    @Override
    public String _getSignatureString(String applicationSecret, String callBackURL, String url)
    {
        String urlParam = "";
        if (url != null && !"".equals(url))
        {
            urlParam = ParameterUtility.nullCheck(url)+"|";
        }		

        String callBackURLParam = "";
        if (callBackURL != null && !"".equals(callBackURL))
        {
            callBackURLParam = ParameterUtility.nullCheck(callBackURL);
        }				

        return ParameterUtility.nullCheck(applicationSecret) + "|" + urlParam + callBackURLParam
        + "|" + format.getValue() + "|" + outputHeight + "|" + outputWidth + "|" + browserHeight 
        + "|" + browserWidth + "|" + getCustomId() + "|" + delay + "|" + targetElement + "|" + customWaterMarkId 
        + "|" + requestAs.getValue() + "|" + getCountry().getValue() + "|" + quality + "|" + hideElement
        + "|" + getExportURL() + "|" + waitForElement + "|" + ParameterUtility.toInt(transparent) + "|" + getEncryptionKey()
        + "|" + ParameterUtility.toInt(noAds) + "|" + post + "|" + getProxy() + "|" + address + "|" + ParameterUtility.toInt(noCookieNotifications)
        + "|" + ParameterUtility.toInt(hd) + "|" + clickElement;
    }    
    
    @Override
    public HashMap<String, String>  _getParameters(String applicationKey, String sig, String callBackURL, String dataName, String dataValue) throws UnsupportedEncodingException
    {
        HashMap<String, String> params = createParameters(applicationKey, sig, callBackURL, dataName, dataValue);		
        params.put("target", ParameterUtility.encode(ParameterUtility.nullCheck(targetElement)));
        params.put("hide", ParameterUtility.encode(ParameterUtility.nullCheck(hideElement)));
        params.put("waitfor", ParameterUtility.encode(ParameterUtility.nullCheck(waitForElement)));
        params.put("requestmobileversion", String.valueOf(requestAs.getValue()));
	params.put("width", String.valueOf(outputWidth));
	params.put("height", String.valueOf(outputHeight));
	params.put("format", format.getValue());
	params.put("bwidth", String.valueOf(browserWidth));
	params.put("customwatermarkid", ParameterUtility.encode(ParameterUtility.nullCheck(customWaterMarkId)));
	params.put("bheight", String.valueOf(browserHeight));
	params.put("delay", String.valueOf(delay));
	params.put("quality", String.valueOf(quality));
        params.put("transparent", String.valueOf(ParameterUtility.toInt(transparent)));
        params.put("noads", String.valueOf(ParameterUtility.toInt(noAds)));
        params.put("post", ParameterUtility.encode(ParameterUtility.nullCheck(post)));        
        params.put("nonotify", String.valueOf(ParameterUtility.toInt(noCookieNotifications)));
        params.put("address", ParameterUtility.encode(ParameterUtility.nullCheck(address)));
        params.put("hd", String.valueOf(ParameterUtility.toInt(hd)));
        params.put("click", ParameterUtility.encode(ParameterUtility.nullCheck(clickElement)));
        
        return params;
    }    
}